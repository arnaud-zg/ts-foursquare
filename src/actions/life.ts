import { NLife } from '../../types/life'
import { lifeReducer } from 'reducers/life'

export enum EActionRequest {
  GET_LIFE = 'GET_LIFE',
}

export const getLife = (): Promise<NLife.ILife> =>
  new Promise(resolve => {
    resolve(
      lifeReducer({
        type: EActionRequest.GET_LIFE,
        payload: {
          life: true,
        },
      })
    )
  })
