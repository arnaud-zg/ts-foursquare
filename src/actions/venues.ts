import { ActionType, createAsyncAction } from 'typesafe-actions'
import {
  NLikes,
  NMiniVenue,
  NRecommendedPlaces,
  NRequest,
  NVenue,
  NVenuesCategories,
} from '../../types'

export const getVenuesCategoriesAsync = createAsyncAction(
  'GET_VENUES_CATEGORIES_REQUEST',
  'GET_VENUES_CATEGORIES_SUCCESS',
  'GET_VENUES_CATEGORIES_FAILURE',
  'GET_VENUES_CATEGORIES_CANCEL'
)<void, NVenuesCategories.ICategory[], Error, string>()

export const getVenuesExploreAsync = createAsyncAction(
  'GET_VENUES_EXPLORE_REQUEST',
  'GET_VENUES_EXPLORE_SUCCESS',
  'GET_VENUES_EXPLORE_FAILURE',
  'GET_VENUES_EXPLORE_CANCEL'
)<
  NRequest.TVenuesExplorePayload,
  NRecommendedPlaces.IGroupItem[],
  Error,
  string
>()

export const getVenuesLikesAsync = createAsyncAction(
  'GET_VENUES_LIKES_REQUEST',
  'GET_VENUES_LIKES_SUCCESS',
  'GET_VENUES_LIKES_FAILURE',
  'GET_VENUES_LIKES_CANCEL'
)<NRequest.IVenuesLikesPayload, NLikes.ILikes, Error, string>()

export const getVenuesNextVenuesAsync = createAsyncAction(
  'GET_VENUES_NEXT_VENUES_REQUEST',
  'GET_VENUES_NEXT_VENUES_SUCCESS',
  'GET_VENUES_NEXT_VENUES_FAILURE',
  'GET_VENUES_NEXT_VENUES_CANCEL'
)<NRequest.IVenuesNextVenuesPayload, NVenue.IVenue[], Error, string>()

export const getVenuesSearchAsync = createAsyncAction(
  'GET_VENUES_SEARCH_REQUEST',
  'GET_VENUES_SEARCH_SUCCESS',
  'GET_VENUES_SEARCH_FAILURE',
  'GET_VENUES_SEARCH_CANCEL'
)<NRequest.TVenuesSearchPayload, NVenue.IVenue[], Error, string>()

export const getVenuesSimilarAsync = createAsyncAction(
  'GET_VENUES_SIMILAR_REQUEST',
  'GET_VENUES_SIMILAR_SUCCESS',
  'GET_VENUES_SIMILAR_FAILURE',
  'GET_VENUES_SIMILAR_CANCEL'
)<NRequest.IVenuesSimilarPayload, NVenue.IVenue[], Error, string>()

export const getVenuesTrendingAsync = createAsyncAction(
  'GET_VENUES_TRENDING_REQUEST',
  'GET_VENUES_TRENDING_SUCCESS',
  'GET_VENUES_TRENDING_FAILURE',
  'GET_VENUES_TRENDING_CANCEL'
)<NRequest.TVenuesTrendingPayload, NVenue.IVenue[], Error, string>()

export const getVenuesSuggestCompletionAsync = createAsyncAction(
  'GET_VENUES_SUGGEST_COMPLETION_REQUEST',
  'GET_VENUES_SUGGEST_COMPLETION_SUCCESS',
  'GET_VENUES_SUGGEST_COMPLETION_FAILURE',
  'GET_VENUES_SUGGEST_COMPLETION_CANCEL'
)<
  NRequest.TVenuesSuggestCompletionPayload,
  NMiniVenue.IMiniVenue[],
  Error,
  string
>()

export const venuesActions = {
  getVenuesCategoriesAsync,
  getVenuesExploreAsync,
  getVenuesLikesAsync,
  getVenuesNextVenuesAsync,
  getVenuesSearchAsync,
  getVenuesSimilarAsync,
  getVenuesSuggestCompletionAsync,
  getVenuesTrendingAsync,
}

export type TVenuesAction = ActionType<typeof venuesActions>
