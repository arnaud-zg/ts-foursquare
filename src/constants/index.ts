export const PROJECT_NAME = 'ts-foursquare'
