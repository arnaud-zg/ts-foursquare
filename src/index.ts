require('isomorphic-fetch')

export * from './standalone'
export * from './utils'

export * from './DEPRECATED_actions'
export * from './DEPRECATED_reducers'
export * from './DEPRECATED_selectors'
