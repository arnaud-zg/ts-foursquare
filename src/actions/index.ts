import { TLifeAction } from './life'
import { TVenuesAction } from './venues'

export type TRootActionType = TVenuesAction | TLifeAction

export type TRootAction = TVenuesAction | TLifeAction

export * from './life'
export * from './venues'
