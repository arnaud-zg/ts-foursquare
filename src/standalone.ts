import { actions } from './actions'

export enum EReturnType {
  OBSERVABLE = 'OBSERVABLE',
  PROMISE = 'PROMISE',
}

export interface IConfigParams {
  config: IStandaloneConfig
}

export interface IStandaloneConfig {
  clientId: string
  clientSecret: string
  accessToken?: string
  returnType?: EReturnType
}

export class tsFoursquare {
  getLists: ReturnType<typeof actions.getLists>
  getPhotosDetails: ReturnType<typeof actions.getPhotosDetails>
  getVenuesCategories: ReturnType<typeof actions.getVenuesCategories>
  getVenuesExplore: ReturnType<typeof actions.getVenuesExplore>
  getVenuesLikes: ReturnType<typeof actions.getVenuesLikes>
  getVenuesListed: ReturnType<typeof actions.getVenuesListed>
  getVenuesNextVenues: ReturnType<typeof actions.getVenuesNextVenues>
  getVenuesSearch: ReturnType<typeof actions.getVenuesSearch>
  getVenuesSimilar: ReturnType<typeof actions.getVenuesSimilar>
  getVenuesSuggestCompletion: ReturnType<
    typeof actions.getVenuesSuggestCompletion
  >
  getVenuesTrending: ReturnType<typeof actions.getVenuesTrending>

  constructor({
    clientId,
    clientSecret,
    returnType = EReturnType.PROMISE,
  }: IStandaloneConfig) {
    const config = {
      clientId,
      clientSecret,
      returnType,
    }

    this.getLists = actions.getLists({ config })
    this.getPhotosDetails = actions.getPhotosDetails({ config })
    this.getVenuesCategories = actions.getVenuesCategories({ config })
    this.getVenuesExplore = actions.getVenuesExplore({ config })
    this.getVenuesLikes = actions.getVenuesLikes({ config })
    this.getVenuesListed = actions.getVenuesListed({ config })
    this.getVenuesNextVenues = actions.getVenuesNextVenues({ config })
    this.getVenuesSearch = actions.getVenuesSearch({ config })
    this.getVenuesSimilar = actions.getVenuesSimilar({ config })
    this.getVenuesSuggestCompletion = actions.getVenuesSuggestCompletion({
      config,
    })
    this.getVenuesTrending = actions.getVenuesTrending({ config })
  }
}
