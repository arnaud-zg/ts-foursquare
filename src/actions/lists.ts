import { NRequest } from '../../types'
import { getObservableLists } from '../services/lists'
import { IStandaloneConfig } from '../standalone'

export interface GetListsProps {
  config?: IStandaloneConfig
  payload: NRequest.IListsPayload
}

export const getLists = ({ config, payload }: GetListsProps) =>
  getObservableLists({ config, payload }).toPromise()
