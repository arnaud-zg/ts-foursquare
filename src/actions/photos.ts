import { NRequest } from '../../types'
import { getObservablePhotosDetails } from '../services/photos'
import { IStandaloneConfig } from '../standalone'

export interface GetPhotosDetailsProps {
  config?: IStandaloneConfig
  payload: NRequest.IPhotosDetailsPayload
}

export const getPhotosDetails = ({ config, payload }: GetPhotosDetailsProps) =>
  getObservablePhotosDetails({ config, payload }).toPromise()
