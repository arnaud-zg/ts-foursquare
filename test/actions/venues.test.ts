import { tsFoursquare } from '../../src/standalone'
import {
  mockingFailingNotFoundFetch,
  mockingFailingNotValidFetch,
  mockingFetch,
} from '../../src/utils/test'
import { payload as payloadGetVenuesCategories } from './__mocks__/getVenuesCategories.resolve'
import { payload as payloadGetVenuesExplore } from './__mocks__/getVenuesExplore.resolve'
import { payload as payloadGetVenuesLikes } from './__mocks__/getVenuesLikes.resolve'
import { payload as payloadGetVenuesListed } from './__mocks__/getVenuesListed.resolve'
import { payload as payloadGetVenuesNextVenues } from './__mocks__/getVenuesNextVenues.resolve'
import { payload as payloadGetVenuesSearch } from './__mocks__/getVenuesSearch.resolve'
import { payload as payloadGetVenuesSimilar } from './__mocks__/getVenuesSimilar.resolve'
import { payload as payloadGetVenuesSuggestCompletion } from './__mocks__/getVenuesSuggestCompletion.resolve'
import { payload as payloadGetVenuesTrending } from './__mocks__/getVenuesTrending.resolve'
import { TActionsKey } from '../../src/actions'

describe('epics/venues', () => {
  test.each`
    scenario                                                             | actionName                      | payloadRequest                                                                                                                               | payloadResponse
    ${'should get venues by location'}                                   | ${'getVenuesSearch'}            | ${{ ll: '40.7099,-73.9622' }}                                                                                                                | ${payloadGetVenuesSearch}
    ${'should get venues by location with intent match parameter'}       | ${'getVenuesSearch'}            | ${{ intent: 'match', ll: '40.7099,-73.9622' }}                                                                                               | ${payloadGetVenuesSearch}
    ${'should get venues by query'}                                      | ${'getVenuesSearch'}            | ${{ query: 'ShopMart' }}                                                                                                                     | ${payloadGetVenuesSearch}
    ${'should get venues near to a location'}                            | ${'getVenuesSearch'}            | ${{ near: 'Chicago, IL' }}                                                                                                                   | ${payloadGetVenuesSearch}
    ${'should get venues'}                                               | ${'getVenuesSearch'}            | ${{ intent: 'browse', ll: '40.7099,-73.9622', query: 'ShopMart', radius: 500 }}                                                              | ${payloadGetVenuesSearch}
    ${'should get similar venues'}                                       | ${'getVenuesSimilar'}           | ${{ venueId: '49b6e8d2f964a52016531fe3' }}                                                                                                   | ${payloadGetVenuesSimilar}
    ${'should get recommended places by location'}                       | ${'getVenuesExplore'}           | ${{ ll: '40.7099,-73.9622' }}                                                                                                                | ${payloadGetVenuesExplore}
    ${'should get recommended places near to a location'}                | ${'getVenuesExplore'}           | ${{ near: 'Chicago, IL' }}                                                                                                                   | ${payloadGetVenuesExplore}
    ${'should get recommended places by location with custom parameter'} | ${'getVenuesExplore'}           | ${{ friendVisits: 'visited', ll: '40.7099,-73.9622', openNow: 1, price: '0,20', section: 'drinks', sortByDistance: 1, query: 'Bubble Tea' }} | ${payloadGetVenuesExplore}
    ${'should get trending venues by location'}                          | ${'getVenuesTrending'}          | ${{ ll: '40.7099,-73.9622' }}                                                                                                                | ${payloadGetVenuesTrending}
    ${'should get trending venues near to a location'}                   | ${'getVenuesTrending'}          | ${{ near: 'Chicago, IL' }}                                                                                                                   | ${payloadGetVenuesTrending}
    ${'should suggest completion for venues by location'}                | ${'getVenuesSuggestCompletion'} | ${{ ll: '40.7099,-73.9622', query: 'Burger' }}                                                                                               | ${payloadGetVenuesSuggestCompletion}
    ${'should suggest completion for venues near to a location'}         | ${'getVenuesSuggestCompletion'} | ${{ near: 'Chicago, IL', query: 'Burger' }}                                                                                                  | ${payloadGetVenuesSuggestCompletion}
    ${'should get venues likes'}                                         | ${'getVenuesLikes'}             | ${{ venueId: '49b6e8d2f964a52016531fe3' }}                                                                                                   | ${payloadGetVenuesLikes}
    ${'should get venues categories'}                                    | ${'getVenuesCategories'}        | ${undefined}                                                                                                                                 | ${payloadGetVenuesCategories}
    ${'should get venues next venues'}                                   | ${'getVenuesNextVenues'}        | ${{ venueId: '49b6e8d2f964a52016531fe3' }}                                                                                                   | ${payloadGetVenuesNextVenues}
    ${'should get a list that this venue appears on'}                    | ${'getVenuesListed'}            | ${{ venueId: '49b6e8d2f964a52016531fe3' }}                                                                                                   | ${payloadGetVenuesListed}
  `(
    '$scenario',
    async ({
      actionName,
      payloadRequest,
      payloadResponse,
    }: {
      actionName: TActionsKey
      payloadRequest: any
      payloadResponse: any
    }) => {
      mockingFetch({ response: { ...payloadResponse } })

      const ts4Sq = new tsFoursquare({ clientId: '', clientSecret: '' })
      const response = await ts4Sq.actions[actionName]({
        payload: payloadRequest,
      })

      expect(response).toMatchSnapshot()
    }
  )
})

describe('epics/venues | error case: no network', () => {
  beforeAll(() => {
    mockingFailingNotFoundFetch()
  })

  test.each`
    scenario                                                     | payloadRequest
    ${'should get venues by location'}                           | ${{ ll: '40.7099,-73.9622' }}
    ${'should get similar venues'}                               | ${{ venueId: '49b6e8d2f964a52016531fe3' }}
    ${'should get recommended places by location'}               | ${{ ll: '40.7099,-73.9622' }}
    ${'should get trending venues by location'}                  | ${{ ll: '40.7099,-73.9622' }}
    ${'should suggest completion for venues by location'}        | ${{ ll: '40.7099,-73.9622', query: 'Burger' }}
    ${'should suggest completion for venues near to a location'} | ${{ near: 'Chicago, IL', query: 'Burger' }}
    ${'should get venues likes'}                                 | ${{ venueId: '49b6e8d2f964a52016531fe3' }}
    ${'should get venues categories'}                            | ${undefined}
    ${'should get venues next venues'}                           | ${{ venueId: '49b6e8d2f964a52016531fe3' }}
    ${'should get a list that this venue appears on'}            | ${{ venueId: '49b6e8d2f964a52016531fe3' }}
  `('$scenario', async ({ payloadRequest }) => {
    const ts4Sq = new tsFoursquare({ clientId: '', clientSecret: '' })
    try {
      await ts4Sq.actions.getPhotosDetails({ payload: payloadRequest })
    } catch (error) {
      expect(error).toMatchSnapshot()
    }
  })
})

describe('epics/venues | error case: not valid response', () => {
  beforeAll(() => {
    mockingFailingNotValidFetch()
  })

  test.each`
    scenario                                                     | payloadRequest
    ${'should get venues by location'}                           | ${{ ll: '40.7099,-73.9622' }}
    ${'should get similar venues'}                               | ${{ venueId: '49b6e8d2f964a52016531fe3' }}
    ${'should get recommended places by location'}               | ${{ ll: '40.7099,-73.9622' }}
    ${'should get trending venues by location'}                  | ${{ ll: '40.7099,-73.9622' }}
    ${'should suggest completion for venues by location'}        | ${{ ll: '40.7099,-73.9622', query: 'Burger' }}
    ${'should suggest completion for venues near to a location'} | ${{ near: 'Chicago, IL', query: 'Burger' }}
    ${'should get venues likes'}                                 | ${{ venueId: '49b6e8d2f964a52016531fe3' }}
    ${'should get venues categories'}                            | ${undefined}
    ${'should get venues next venues'}                           | ${{ venueId: '49b6e8d2f964a52016531fe3' }}
    ${'should get a list that this venue appears on'}            | ${{ venueId: '49b6e8d2f964a52016531fe3' }}
  `('$scenario', async ({ payloadRequest }) => {
    const ts4Sq = new tsFoursquare({ clientId: '', clientSecret: '' })
    try {
      await ts4Sq.actions.getPhotosDetails({ payload: payloadRequest })
    } catch (error) {
      expect(error).toMatchSnapshot()
    }
  })
})
