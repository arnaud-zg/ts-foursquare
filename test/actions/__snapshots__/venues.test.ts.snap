// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions.venues should create an action of type REQUEST_GET_VENUES_SEARCH 1`] = `
Object {
  "error": undefined,
  "meta": undefined,
  "payload": Object {
    "ll": "40.7099,-73.9622",
  },
  "type": "REQUEST_GET_VENUES_SEARCH",
}
`;

exports[`actions.venues should create an action of type REQUEST_GET_VENUES_SEARCH 2`] = `
Object {
  "error": undefined,
  "meta": undefined,
  "payload": Object {
    "intent": "match",
    "ll": "40.7099,-73.9622",
  },
  "type": "REQUEST_GET_VENUES_SEARCH",
}
`;

exports[`actions.venues should create an action of type REQUEST_GET_VENUES_SEARCH 3`] = `
Object {
  "error": undefined,
  "meta": undefined,
  "payload": Object {
    "query": "ShopMart",
  },
  "type": "REQUEST_GET_VENUES_SEARCH",
}
`;

exports[`actions.venues should create an action of type REQUEST_GET_VENUES_SEARCH 4`] = `
Object {
  "error": undefined,
  "meta": undefined,
  "payload": Object {
    "near": "Chicago, IL",
  },
  "type": "REQUEST_GET_VENUES_SEARCH",
}
`;

exports[`actions.venues should create an action of type REQUEST_GET_VENUES_SEARCH 5`] = `
Object {
  "error": undefined,
  "meta": undefined,
  "payload": Object {
    "intent": "browse",
    "ll": "40.7099,-73.9622",
    "query": "ShopMart",
    "radius": 500,
  },
  "type": "REQUEST_GET_VENUES_SEARCH",
}
`;

exports[`actions.venues should create an action of type RESOLVE_GET_VENUES_SEARCH 1`] = `
Object {
  "error": undefined,
  "meta": undefined,
  "payload": Array [],
  "type": "RESOLVE_GET_VENUES_SEARCH",
}
`;
